app/Models/Forum/Topic.php

```diff
@@ -126,25 +129,35 @@ public function addPost($poster, $body)
 
     public function removePost($post, $user = null)
     {
-        DB::transaction(function () use ($post, $user) {
-            $post->delete();
+        $this->validationErrors()->reset();
 
-            if ($this->posts()->exists() === true) {
-                $this->refreshCache();
-            } else {
-                $this->delete();
-            }
+        try {
+            return DB::transaction(function () use ($post, $user) {
+                if ($post->delete() === false) {
+                    $this->validationErrors()->addTranslated('post', $post->validationErrors()->toSentence());
 
-            if ($this->forum !== null) {
-                $this->forum->refreshCache();
-            }
+                    throw new ModelNotSavedException('failed deleting post');
+                }
 
-            if ($post->user !== null) {
-                $post->user->refreshForumCache($this->forum, -1);
-            }
-        });
+                if ($this->posts()->exists() === true) {
+                    $this->refreshCache();
+                } else {
+                    $this->delete();
+                }
 
-        return true;
+                if ($this->forum !== null) {
+                    $this->forum->refreshCache();
+                }
+
+                if ($post->user !== null) {
+                    $post->user->refreshForumCache($this->forum, -1);
+                }
+
+                return true;
+            });
+        } catch (ModelNotSavedException $_e) {
```

Comment:
may be better to just let the exception propagate and let the caller handle it instead of
`if false throw catch return false if false throw catch return false` party